File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/files.clj
(ns hydra.overlay.clojure.lib.files
  (:require [hydra.file :refer :all]
            [hydra.time :refer :all])
  (:import [java.io IOException FileNotFoundException]
           [java.nio.file Files Paths LinkOption OpenOption StandardOpenOption StandardCopyOption
            AccessDeniedException FileAlreadyExistsException NoSuchFileException
            NotDirectoryException InvalidPathException FileVisitResult SimpleFileVisitor]
           [java.nio.file.attribute BasicFileAttributes]))

;; Clojure implementations of hydra.lib.files primitives (#494).
;;
;; The Hydra type effect<t> is transparent in the Lisp dialects (effect<t> = t), so these
;; helpers perform real java.nio file I/O eagerly and return a Hydra Either<FileError, T>.
;; A recoverable file-system failure becomes (list :left error); success becomes (list :right value).
;; Mirrors the withFileError/classify pattern of the Haskell reference (Hydra.Haskell.Lib.Files).
;;
;; Runtime representations (Clojure target):
;;   Either        : (list :left v) | (list :right v)
;;   FileError     : (list :<variant> payload), where <variant> is one of
;;                   :already_exists | :invalid_path | :not_found | :other | :permission_denied
;;                   and payload is a FilePath (string) for path-carrying variants, a string for the rest.
;;   FilePath      : a bare string (the FilePath wrap is transparent at the value level).
;;   unit          : nil
;;   binary        : a Clojure vector/seq of byte ints (each 0..255), matching the
;;                   Clojure target's binary representation (see hydra.overlay.clojure.lib.literals).
;;                   Convert to/from a Java byte-array only for the actual java.nio I/O.
;;
;; All functions are curried, matching the Clojure prim runtime style.

;; ---- Helpers (not primitives) ----

;; binary (vector of ints 0..255) -> Java byte-array, for I/O.
(defn- binary->bytes [contents]
  (byte-array (map unchecked-byte contents)))

;; Java byte-array -> binary (vector of unsigned ints 0..255).
(defn- bytes->binary [^bytes ba]
  (vec (map #(bit-and % 0xff) ba)))

(defn- message [^Exception e]
  (let [m (.getMessage e)]
    (if (nil? m) (.getSimpleName (class e)) m)))

;; Classify an IOException into a FileError, mirroring the Haskell/Scala host's classify.
(defn- classify [path ^IOException e]
  (cond
    (instance? FileAlreadyExistsException e) (list :already_exists path)
    (instance? NoSuchFileException e)        (list :not_found path)
    (instance? FileNotFoundException e)      (list :not_found path)
    (instance? NotDirectoryException e)      (list :not_found path)
    (instance? AccessDeniedException e)      (list :permission_denied path)
    :else                                    (list :other (message e))))

;; Run a file-system action, translating any IOException into the appropriate FileError.
(defn- with-file-error [path action]
  (try
    (list :right (action))
    (catch InvalidPathException e (list :left (list :invalid_path (message e))))
    (catch IOException e          (list :left (classify path e)))
    (catch Exception e            (list :left (list :other (message e))))))

(defn- path-of [s] (Paths/get s (into-array String [])))

;; Recursively copy a directory tree, mirroring the Haskell host's copyDirectoryRecursive.
(defn- copy-directory-recursive [source destination]
  (Files/walkFileTree source
    (proxy [SimpleFileVisitor] []
      (preVisitDirectory [dir attrs]
        (Files/createDirectories (.resolve destination (.relativize source dir))
          (into-array java.nio.file.attribute.FileAttribute []))
        FileVisitResult/CONTINUE)
      (visitFile [file attrs]
        (Files/copy file (.resolve destination (.relativize source file))
          (into-array java.nio.file.CopyOption [StandardCopyOption/REPLACE_EXISTING]))
        FileVisitResult/CONTINUE))))

;; Recursively remove a directory tree, mirroring the Haskell host's removeDirectoryRecursive.
(defn- remove-directory-recursive [path]
  (Files/walkFileTree path
    (proxy [SimpleFileVisitor] []
      (visitFile [file attrs]
        (Files/delete file)
        FileVisitResult/CONTINUE)
      (postVisitDirectory [dir exc]
        (Files/delete dir)
        FileVisitResult/CONTINUE))))

;; ---- Primitives ----

;; appendFile :: FilePath -> binary -> effect<Either<FileError, unit>>
(def hydra_lib_files_append_file
  "Append raw bytes to the end of a file, creating it if absent."
  (fn [path] (fn [contents]
    (with-file-error path
      (fn []
        (Files/write (path-of path) (binary->bytes contents)
          (into-array OpenOption [StandardOpenOption/CREATE StandardOpenOption/APPEND]))
        nil)))))

;; copy :: Bool -> FilePath -> FilePath -> effect<Either<FileError, unit>>
(def hydra_lib_files_copy
  "Copy source to destination; when recursive, source may be a directory whose tree is copied."
  (fn [recursive] (fn [source] (fn [destination]
    (with-file-error source
      (fn []
        (let [source-path (path-of source)
              destination-path (path-of destination)
              source-is-dir (Files/isDirectory source-path (into-array LinkOption []))]
          (cond
            (and recursive source-is-dir) (copy-directory-recursive source-path destination-path)
            source-is-dir (throw (java.nio.file.FileSystemException. source nil "is a directory, but recursive is false"))
            :else (Files/copy source-path destination-path
                    (into-array java.nio.file.CopyOption [StandardCopyOption/REPLACE_EXISTING])))
          nil)))))))

;; createDirectory :: Bool -> FilePath -> effect<Either<FileError, unit>>
(def hydra_lib_files_create_directory
  "Create a directory; when recursive, create missing parents (mkdir -p)."
  (fn [recursive] (fn [path]
    (with-file-error path
      (fn []
        (let [p (path-of path)]
          (if recursive
            (Files/createDirectories p (into-array java.nio.file.attribute.FileAttribute []))
            (Files/createDirectory p (into-array java.nio.file.attribute.FileAttribute []))))
        nil)))))

;; exists :: FilePath -> effect<Either<FileError, Bool>>
(def hydra_lib_files_exists
  "Test whether anything exists at the given path (no error on absence)."
  (fn [path]
    (with-file-error path
      (fn [] (Files/exists (path-of path) (into-array LinkOption []))))))

;; listDirectory :: FilePath -> effect<Either<FileError, [FilePath]>>
;; Returns the bare entry names (not full paths), mirroring the Haskell listDirectory.
(def hydra_lib_files_list_directory
  "List the immediate entries of a directory as bare names."
  (fn [path]
    (with-file-error path
      (fn []
        (let [stream (Files/list (path-of path))]
          (try
            (->> (.iterator stream)
                 iterator-seq
                 (mapv (fn [p]
                         (let [n (.getFileName p)]
                           (if (nil? n) (.toString p) (.toString n))))))
            (finally (.close stream))))))))

;; readFile :: FilePath -> effect<Either<FileError, binary>>
(def hydra_lib_files_read_file
  "Read the entire contents of a file as raw bytes."
  (fn [path]
    (with-file-error path
      (fn [] (bytes->binary (Files/readAllBytes (path-of path)))))))

;; removeDirectory :: Bool -> FilePath -> effect<Either<FileError, unit>>
(def hydra_lib_files_remove_directory
  "Remove a directory; when recursive, remove its entire contents (rm -r)."
  (fn [recursive] (fn [path]
    (with-file-error path
      (fn []
        (if recursive
          (remove-directory-recursive (path-of path))
          (Files/delete (path-of path)))
        nil)))))

;; removeFile :: FilePath -> effect<Either<FileError, unit>>
(def hydra_lib_files_remove_file
  "Remove a file (POSIX unlink)."
  (fn [path]
    (with-file-error path
      (fn [] (Files/delete (path-of path)) nil))))

;; rename :: FilePath -> FilePath -> effect<Either<FileError, unit>>
(def hydra_lib_files_rename
  "Rename or move a file or directory from source to destination."
  (fn [source] (fn [destination]
    (with-file-error source
      (fn []
        (Files/move (path-of source) (path-of destination)
          (into-array java.nio.file.CopyOption []))
        nil)))))

;; Classify BasicFileAttributes into a hydra.file.FileType union value.
(defn- file-type [^BasicFileAttributes attrs]
  (cond
    (.isDirectory attrs)     (list :directory nil)
    (.isSymbolicLink attrs)  (list :link nil)
    (.isRegularFile attrs)   (list :regular nil)
    :else                    (list :regular nil)))

;; java.time.Instant -> hydra.time.Timespec (a hydra_time_timespec defrecord {:seconds :nanoseconds}).
(defn- timespec [^java.time.Instant instant]
  (->hydra_time_timespec (.getEpochSecond instant) (long (.getNano instant))))

;; status :: FilePath -> effect<Either<FileError, FileStatus>>
;; Retrieve metadata about the file at path (POSIX stat). Symbolic links are followed.
;; FileStatus is a hydra_file_file_status defrecord
;; {:file_type :size :modification_time :access_time :status_change_time}.
(def hydra_lib_files_status
  "Retrieve metadata about the file at path (POSIX stat)."
  (fn [path]
    (with-file-error path
      (fn []
        (let [attrs (Files/readAttributes (path-of path) BasicFileAttributes (into-array LinkOption []))]
          (->hydra_file_file_status
            (file-type attrs)
            (.size attrs)
            (timespec (.toInstant (.lastModifiedTime attrs)))
            (list :given (timespec (.toInstant (.lastAccessTime attrs))))
            (list :none)))))))

;; writeFile :: FilePath -> binary -> effect<Either<FileError, unit>>
(def hydra_lib_files_write_file
  "Replace the file at path with the raw bytes contents, creating it if necessary."
  (fn [path] (fn [contents]
    (with-file-error path
      (fn []
        (Files/write (path-of path) (binary->bytes contents) (into-array OpenOption []))
        nil)))))

File: overlay/scala/hydra-kernel/src/main/scala/hydra/overlay/scala/lib/files.scala
package hydra.overlay.scala.lib

import hydra.error.file.FileError

import java.io.{FileNotFoundException, IOException}
import java.nio.file.{AccessDeniedException, FileAlreadyExistsException, Files, InvalidPathException,
  NoSuchFileException, Paths, StandardOpenOption}

/**
 * Scala implementations of hydra.lib.files primitives (#494).
 *
 * The effect type is transparent in Scala (effect<t> = t), so these helpers perform real
 * java.nio file I/O eagerly and return scala.util.Either[FileError, T]. A recoverable
 * file-system failure becomes Left(error); success becomes Right(value). Mirrors the
 * withFileError/classify pattern of the Haskell and Java hosts.
 *
 * hydra.file.FilePath is a wrapped string, which the Scala coder renders as a plain
 * String alias, so a path argument is used directly. The Hydra `binary` type maps to a
 * base64-encoded String in the Scala host (see literals.binaryToBytes), so binary file
 * contents arrive/leave as base64 Strings and are base64-decoded/encoded around the raw
 * byte I/O. hydra.core.unit maps to Scala Unit, so mutating operations succeed with Right(()).
 */
object files:
  /** Append binary contents (base64-encoded) to the end of a file, creating it if absent. */
  def appendFile(path: String)(contents: String): Either[FileError, Unit] =
    withFileError(path) {
      Files.write(Paths.get(path), java.util.Base64.getDecoder.decode(contents),
        StandardOpenOption.CREATE, StandardOpenOption.APPEND)
      ()
    }

  /** Copy source to destination; when recursive, source may be a directory whose tree is copied. */
  def copy(recursive: Boolean)(source: String)(destination: String): Either[FileError, Unit] =
    withFileError(source) {
      val sourcePath = Paths.get(source)
      val destinationPath = Paths.get(destination)
      if recursive && Files.isDirectory(sourcePath) then copyDirectoryRecursive(sourcePath, destinationPath)
      else {
        if Files.isDirectory(sourcePath) then
          throw new _root_.java.nio.file.FileSystemException(source, null, "is a directory, but recursive is false")
        Files.copy(sourcePath, destinationPath, _root_.java.nio.file.StandardCopyOption.REPLACE_EXISTING)
      }
      ()
    }

  /** Create a directory; when recursive, create missing parents (mkdir -p). */
  def createDirectory(recursive: Boolean)(path: String): Either[FileError, Unit] =
    withFileError(path) {
      val p = Paths.get(path)
      if recursive then Files.createDirectories(p) else Files.createDirectory(p)
      ()
    }

  /** Create a symbolic link at link, pointing to target (stored verbatim). No force flag: an
    * occupied link path (including a dangling symlink) fails with alreadyExists. */
  def createSymlink(target: String)(link: String): Either[FileError, Unit] =
    withFileError(link) {
      Files.createSymbolicLink(Paths.get(link), Paths.get(target))
      ()
    }

  /** Test whether a path exists (no error on absence). */
  def exists(path: String): Either[FileError, Boolean] =
    withFileError(path) { Files.exists(Paths.get(path)) }

  /** List the immediate entries of a directory as bare names (FilePath = String). */
  def listDirectory(path: String): Either[FileError, Seq[String]] =
    withFileError(path) {
      val stream = Files.list(Paths.get(path))
      try {
        val buf = _root_.scala.collection.mutable.ArrayBuffer.empty[String]
        stream.forEach { p =>
          val name = p.getFileName
          buf += (if name == null then p.toString else name.toString)
        }
        buf.toSeq
      } finally stream.close()
    }

  /** Read the entire contents of a file, returning the bytes as a base64-encoded String. */
  def readFile(path: String): Either[FileError, String] =
    withFileError(path) {
      java.util.Base64.getEncoder.encodeToString(Files.readAllBytes(Paths.get(path)))
    }

  /** Read the target of the symbolic link at path, verbatim and unresolved (relative stays
    * relative; a dangling target is returned as-is). Fails with invalidPath if path is not a
    * symbolic link. */
  def readSymlink(path: String): Either[FileError, String] =
    withFileError(path) {
      val p = Paths.get(path)
      if !Files.exists(p, _root_.java.nio.file.LinkOption.NOFOLLOW_LINKS) then
        throw new NoSuchFileException(path)
      if !Files.isSymbolicLink(p) then
        throw new InvalidPathException(path, "not a symbolic link")
      Files.readSymbolicLink(p).toString
    }

  /** Remove a directory; when recursive, remove its entire contents (rm -r). */
  def removeDirectory(recursive: Boolean)(path: String): Either[FileError, Unit] =
    withFileError(path) {
      if recursive then removeDirectoryRecursive(Paths.get(path)) else Files.delete(Paths.get(path))
      ()
    }

  /** Remove a file (POSIX unlink). */
  def removeFile(path: String): Either[FileError, Unit] =
    withFileError(path) {
      Files.delete(Paths.get(path))
      ()
    }

  /** Rename or move a file or directory. */
  def rename(source: String)(destination: String): Either[FileError, Unit] =
    withFileError(source) {
      Files.move(Paths.get(source), Paths.get(destination))
      ()
    }

  /** Retrieve metadata about the file at path. When followLinks is true (POSIX stat), a
    * symbolic link's metadata is that of its target, and a dangling link is not found. When
    * false (POSIX lstat), a symbolic link's own metadata is reported (fileType link), and a
    * dangling link is not an error. */
  def status(followLinks: Boolean)(path: String): Either[FileError, hydra.file.FileStatus] =
    withFileError(path) {
      import _root_.java.nio.file.LinkOption
      val options: Array[LinkOption] =
        if followLinks then Array.empty else Array(LinkOption.NOFOLLOW_LINKS)
      val attrs = Files.readAttributes(Paths.get(path), "unix:*", options*)
      val mode = attrs.get("mode").asInstanceOf[Integer].intValue
      val size = attrs.get("size").asInstanceOf[java.lang.Long].longValue
      val mtime = attrs.get("lastModifiedTime").asInstanceOf[_root_.java.nio.file.attribute.FileTime]
      val atime = attrs.get("lastAccessTime").asInstanceOf[_root_.java.nio.file.attribute.FileTime]
      val ctime = attrs.get("ctime").asInstanceOf[_root_.java.nio.file.attribute.FileTime]
      hydra.file.FileStatus(
        fileTypeFromMode(mode),
        size,
        timespec(mtime.toInstant),
        Some(timespec(atime.toInstant)),
        Some(timespec(ctime.toInstant)))
    }

  /** Write binary contents (base64-encoded) as the complete contents of a file. */
  def writeFile(path: String)(contents: String): Either[FileError, Unit] =
    withFileError(path) {
      Files.write(Paths.get(path), java.util.Base64.getDecoder.decode(contents))
      ()
    }

  // ---- Helpers (not primitives) ----

  // POSIX st_mode file-type bits (<sys/stat.h> S_IFMT and the individual S_IF* macros).
  private val S_IFMT   = 0170000
  private val S_IFSOCK = 0140000
  private val S_IFLNK  = 0120000
  private val S_IFREG  = 0100000
  private val S_IFBLK  = 0060000
  private val S_IFDIR  = 0040000
  private val S_IFCHR  = 0020000
  private val S_IFIFO  = 0010000

  private def fileTypeFromMode(mode: Int): hydra.file.FileType =
    (mode & S_IFMT) match {
      case S_IFDIR  => hydra.file.FileType.directory
      case S_IFLNK  => hydra.file.FileType.link
      case S_IFBLK  => hydra.file.FileType.block
      case S_IFCHR  => hydra.file.FileType.character
      case S_IFIFO  => hydra.file.FileType.fifo
      case S_IFSOCK => hydra.file.FileType.socket
      case _        => hydra.file.FileType.regular
    }

  private def timespec(instant: _root_.java.time.Instant): hydra.time.Timespec =
    hydra.time.Timespec(instant.getEpochSecond, instant.getNano.toLong)

  private def copyDirectoryRecursive(source: _root_.java.nio.file.Path, destination: _root_.java.nio.file.Path): Unit = {
    import _root_.java.nio.file.{FileVisitResult, SimpleFileVisitor}
    import _root_.java.nio.file.attribute.BasicFileAttributes
    Files.walkFileTree(source, new SimpleFileVisitor[_root_.java.nio.file.Path] {
      override def preVisitDirectory(dir: _root_.java.nio.file.Path, attrs: BasicFileAttributes): FileVisitResult = {
        Files.createDirectories(destination.resolve(source.relativize(dir)))
        FileVisitResult.CONTINUE
      }
      override def visitFile(file: _root_.java.nio.file.Path, attrs: BasicFileAttributes): FileVisitResult = {
        Files.copy(file, destination.resolve(source.relativize(file)), _root_.java.nio.file.StandardCopyOption.REPLACE_EXISTING)
        FileVisitResult.CONTINUE
      }
    })
    ()
  }

  private def removeDirectoryRecursive(path: _root_.java.nio.file.Path): Unit = {
    import _root_.java.nio.file.{FileVisitResult, SimpleFileVisitor}
    import _root_.java.nio.file.attribute.BasicFileAttributes
    Files.walkFileTree(path, new SimpleFileVisitor[_root_.java.nio.file.Path] {
      override def visitFile(file: _root_.java.nio.file.Path, attrs: BasicFileAttributes): FileVisitResult = {
        Files.delete(file)
        FileVisitResult.CONTINUE
      }
      override def postVisitDirectory(dir: _root_.java.nio.file.Path, exc: IOException): FileVisitResult = {
        Files.delete(dir)
        FileVisitResult.CONTINUE
      }
    })
    ()
  }

  /** Run a file-system action, translating any IOException into the appropriate FileError. */
  private def withFileError[T](path: String)(action: => T): Either[FileError, T] =
    try Right(action)
    catch {
      case e: InvalidPathException => Left(FileError.invalidPath(message(e)))
      case e: IOException          => Left(classify(path, e))
      case e: Exception            => Left(FileError.other(message(e)))
    }

  /** Classify an IOException into a FileError, mirroring the Haskell/Java host's classify. */
  private def classify(path: String, e: IOException): FileError = e match {
    case _: FileAlreadyExistsException => FileError.alreadyExists(path)
    case _: NoSuchFileException        => FileError.notFound(path)
    case _: FileNotFoundException      => FileError.notFound(path)
    case _: AccessDeniedException      => FileError.permissionDenied(path)
    case _                             => FileError.other(message(e))
  }

  private def message(e: Exception): String = {
    val msg = e.getMessage
    if msg == null then e.getClass.getSimpleName else msg
  }
